pkg/summarizer/summary.go

```diff
@@ -73,142 +70,131 @@ type gridReader func(ctx context.Context) (io.ReadCloser, time.Time, int64, erro
-					log.WithError(err).Error("Cannot write summary")
-					errCh <- errors.New(dash.Name)
-					continue
-				}
-				log.Info("Wrote dashboard summary")
-				errCh <- nil
-			}
-		}()
+func (cs *configSnapshot) dashboard(name string) *configpb.Dashboard {
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	return cs.dashboards[name]
+}
+
+func (cs *configSnapshot) group(name string) *configpb.TestGroup {
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	return cs.groups[name]
+}
+
+func (cs *configSnapshot) dashboardTestGroups(dashboardName string) (*configpb.Dashboard, map[string]*configpb.TestGroup) {
+	cs.lock.RLock()
+	defer cs.lock.RLock()
+
+	d := cs.dashboards[dashboardName]
+	if d == nil {
+		return nil, nil
+	}
+	gs := make(map[string]*configpb.TestGroup, len(d.DashboardTab))
+	for _, tab := range d.DashboardTab {
+		n := tab.TestGroupName
+		gs[n] = cs.groups[n]
 	}
+	return d, gs
+}
 
-	resultCh := make(chan error)
-	go func() {
-		var errs []string
-		for err := range errCh {
-			if err == nil {
-				mets.Success()
-				continue
-			}
-			if mets.Errors != nil {
-				mets.Error()
-			}
-			errs = append(errs, err.Error())
-		}
-		if n := len(errs); n > 0 {
-			resultCh <- fmt.Errorf("failed to update %d dashboards: %v", n, strings.Join(errs, ", "))
-		}
-		resultCh <- nil
-		close(resultCh)
-	}()
+func (cs *configSnapshot) update(ctx context.Context, log logrus.FieldLogger, client gcs.ConditionalClient, configPath gcs.Path, summaryPathPrefix, dashboard string, q *config.DashboardQueue, freq time.Duration) (*storage.ReaderObjectAttrs, error) {
+	cfg, attrs, err := fetchConfig(ctx, client, configPath, dashboard)
+	if err != nil {
+		return nil, err
+	}
 
-	if dashboard == "" {
-		var err error
-		generations, err = sortDashboards(ctx, log, client, configPath, summaryPathPrefix, cfg.Dashboards)
+	now := time.Now()
+
+	dashboards := cfg.Dashboards
+	paths := make([]gcs.Path, 0, len(dashboards))
+	for _, d := range dashboards {
+		path, err := summaryPath(configPath, summaryPathPrefix, d.Name)
 		if err != nil {
-			log.WithError(err).Warning("Failed to sort dashboards")
+			return nil, fmt.Errorf("bad dashboard path: %s: %w", d.Name, err)
 		}
+		paths = append(paths, *path)
 	}
-	currently := util.Progress(ctx, log, time.Minute, len(cfg.Dashboards), "Summarizing dashboards...")
-	for i, d := range cfg.Dashboards {
-		currently(i)
-		if dashboard != "" && dashboard != d.Name {
-			log.WithField("dashboard", d.Name).Info("Skipping")
-			continue
+
+	stats := gcs.Stat(ctx, client, 10, paths...)
+	whens := make(map[string]time.Time, len(stats))
+	var wg sync.WaitGroup
+	for i, stat := range stats {
+		name := dashboards[i].Name
+		switch {
+		case stat.Attrs != nil:
+			whens[name] = stat.Attrs.Updated.Add(freq)
+		default:
+			if errors.Is(stat.Err, storage.ErrObjectNotExist) {
+				wg.Add(1)
+				defer func(i int) {
+					defer wg.Done()
+					if _, err := lockDashboard(ctx, client, paths[i], 0); err != nil && !gcs.IsPreconditionFailed(err) {
```

Comment:
Should we log here when err != nil? Is there something magic here? This seems identical to `lockDashboard(ctx, client, paths[i], 0)`

Reply:
Done